controllers/kubeadm_control_plane_controller.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=kubeadmcontrolplanes;kubeadmcontrolplanes/status,verbs=get;list;watch;create;update;patch;delete
+
+// KubeadmControlPlaneReconciler reconciles a KubeadmControlPlane object
+type KubeadmControlPlaneReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	controller controller.Controller
+	recorder   record.EventRecorder
+}
+
+func (r *KubeadmControlPlaneReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.KubeadmControlPlane{}).
+		WithOptions(options).
+		Build(r)
+
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.controller = c
+	r.recorder = mgr.GetEventRecorderFor("kubeadm-control-plane-controller")
+
+	return nil
+}
+
+func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Fetch the KubeadmControlPlane instance.
+	kubeadmControlPlane := &clusterv1.KubeadmControlPlane{}
+	if err := r.Client.Get(ctx, req.NamespacedName, kubeadmControlPlane); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(kubeadmControlPlane, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
Do you want to log here? If not, it falls back to logging from a controller-runtime file:line, and it's hard to tell where the error came from.

Also, I feel like this is a call that _should_ never error, but if it does, I'm not sure that returning an error (and thus re-enqueueing) would yield a different result the next time it's reconciled? I checked and our other controllers return the error, but maybe they shouldn't? This isn't a blocker to merging and more me trying to minimize "useless" requeues. Might be worth an issue to make all the controllers consistent, depending on what we decide to do here.

Reply:
Will create an issue based on discussions from San Diego.

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/1840

Reply:
> Also, I feel like this is a call that should never error, but if it does, I'm not sure that returning an error (and thus re-enqueueing) would yield a different result the next time it's reconciled? I checked and our other controllers return the error, but maybe they shouldn't? This isn't a blocker to merging and more me trying to minimize "useless" requeues. Might be worth an issue to make all the controllers consistent, depending on what we decide to do here.

Ideally it would never return an error, but if the api server is briefly unavailable (member behind a load balancer that was routed the request hanging prior to health check to remove from the LB triggering, for example), I think that we would indeed want to requeue to ensure that we record the status prior to the resync period.

Reply:
That said, we could potentially implement an exponential backoff/retry in the patch helper to achieve the same result.

Reply:
Fixed the logging here, we can re-examine how we handle patching failures as a separate issue.